src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -964,6 +1085,64 @@ public DocumentId AddTextContainer(SourceTextContainer textContainer, string ful
                 return documentId;
             }
 
+            public void AddDynamicFile(IDynamicFileInfoProvider fileInfoProvider, DynamicFileInfo fileInfo, ImmutableArray<string> folders)
+            {
+                var documentInfo = CreateDocumentInfoFromFileInfo(fileInfo, folders.NullToEmpty());
+                var documentId = documentInfo.Id;
+
+                lock (_project._gate)
+                {
+                    var filePath = documentInfo.FilePath;
+                    if (_documentPathsToDocumentIds.ContainsKey(filePath))
+                    {
+                        throw new ArgumentException($"'{filePath}' has already been added to this project.", nameof(filePath));
+                    }
+
+                    _documentPathsToDocumentIds.Add(filePath, documentId);
+
+                    _documentIdToDynamicFileInfoProvider.Add(documentId, fileInfoProvider);
+
+                    if (_project._eventSubscriptionTracker.Add(fileInfoProvider))
+                    {
+                        // subscribe to the event when we use this provider the first time
+                        fileInfoProvider.Updated += _project.OnDynamicFileInfoUpdated;
+                    }
+
+                    if (_project._activeBatchScopes > 0)
+                    {
+                        _documentsAddedInBatch.Add(documentInfo);
+                    }
+                    else
+                    {
+                        // right now, assumption is dynamically generated file can never be opened in editor
+                        _project._workspace.ApplyChangeToWorkspace(w => _documentAddAction(w, documentInfo));
+                    }
+                }
+            }
+
+            public IDynamicFileInfoProvider RemoveDynamicFile(string fullPath)
+            {
+                if (string.IsNullOrEmpty(fullPath))
+                {
+                    throw new ArgumentException($"{nameof(fullPath)} isn't a valid path.", nameof(fullPath));
+                }
+
+                lock (_project._gate)
+                {
+                    if (!_documentPathsToDocumentIds.TryGetValue(fullPath, out var documentId) ||
+                        !_documentIdToDynamicFileInfoProvider.TryGetValue(documentId, out var fileInfoProvider))
+                    {
+                        throw new ArgumentException($"'{fullPath}' is not a dynamic file of this project.");
+                    }
+
+                    _documentIdToDynamicFileInfoProvider.Remove(documentId);
+
+                    RemoveFileInternal(documentId, fullPath);
+
+                    return fileInfoProvider;
```

Comment:
💭 It seems a little strange to be returning this, since the provider wouldn't be used after the file is removed